rust/tool/tests/install.rs

```diff
@@ -60,26 +61,28 @@ fn overwrite_unsigned_images() -> Result<()> {
     let output2 = common::lanzaboote_install(0, esp.path(), generation_links)?;
     assert!(output2.status.success());
 
-    assert!(verify_signature(&image1)?);
+    // Assert existing stub was not overwritten.
+    assert!(!verify_signature(&image1)?);
     assert!(verify_signature(&image2)?);
 
     Ok(())
 }
 
 #[test]
-fn overwrite_unsigned_files() -> Result<()> {
+fn do_not_overwrite_files() -> Result<()> {
```

Comment:
In a way it feels like a regression that we now cannot detect changes to the content of (some) files anymore. However I'm fine with it.

Hopefully, some time in the future, when goblin has the capability to read PE files without the signature and we drop support for initrd secrets we can move to a fully content addressed system.

@RaitoBezarius what do you think?

Reply:
This is already mentioned in the drawbacks above. It's part of the trade-off, because detecting all changes obviously requires reading everything at least once, which is very slow.

Reply:
The only concern I have for not overwriting files is that repair guidelines are unclear now.

We need to explicitly ask people to `rm -rf` their ESP and rerun the bootloader install.

You say:

> In this case, it is questionable whether overwriting the files fixes the problem anyway.

But in my experience, I see some folks with accidental file corruption and a way to unstuck them is very necessary.

It is also clear that lanzaboote cannot be the only thing to fix this.

Reply:
I agree that it is good to have repair guidelines, but I am just doubtful that blindly overwriting files on a corrupted filesystem is actually a fix. Do we have some examples of accidental file corruption in the ESP (e.g. what files are commonly affected)? If so, then we could just document to run an fsck, delete the corrupted files, and rerun `nixos-rebuild boot`.

Reply:
Yes, more than often the kernel or initrd hash can be affected because of a corruption in the kernel or the hash.
I am not sure what fsck can do about this, though, FAT32 is not a journaled filesystem. It does not know what is broken or what is it.

Reply:
Yes, but how does that happen? We do atomic writes, hence a kernel or initrd hash mismatch can only result when the filesystem itself is corrupted (short of user error such as `dd if=/dev/urandom of=/boot/EFI/nixos/kernel-6.5.4-cM5y3rGuuVlMAP6NBtgT6c_LTOuipLZ9jolppk79J3Y.efi`). What do you suggest as a fix instead? (This is already relevant without this change btw.)

Reply:
Atomic writes are not achievable on FAT32, power loss at any point during install can cancel the atomic write and this is not recoverable because no journal.

The fix can only be found by using OS recovery measures provided by UEFI and storing a golden ESP partition somewhere or whatever.

Reply:
It's sad that FAT32 cannot even offer the most basic reliability guarantees of atomic writes (I thought it did offer atomic writes even though its rename is not atomic, but apparently it can't even guarantee that). How does this failure usually manifest itself in the cases you have experienced (truncated files, garbage content, …)? I'd like to be able to document this properly.

And sorry if I wasn't clear, when I said "fix" I really meant a somewhat proper sequence of steps that should be performed when the power failed and the system cannot boot due to the corruption experienced. Clearly it seems to be the case that "rewrite all files and hope for the best" seems to be better than nothing. I just want to know if there are better ways so that I can document this as a recovery procedure. Would an fsck help? Or is it best to mkfs over?

Reply:
(Side note: The comment in the `atomic_copy` function should then be changed as well. And I now noticed that its implementation is incorrect, since the sync is missing, but I'm not sure how much that changes on FAT after what you have written above.)

Reply:
> How does this failure usually manifest itself in the cases you have experienced (truncated files, garbage content, …)? I'd like to be able to document this properly.

I didn't dive into failures related to FAT32 brokenness, but apparently, out there, there are in-the-wild broken implementations of FAT32 who does not know how to behave with the dirty flag set and does wrong things.

So globally, IME, anything can happen. Personally, I have seen truncated files mostly.

Reply:
> And sorry if I wasn't clear, when I said "fix" I really meant a somewhat proper sequence of steps that should be performed when the power failed and the system cannot boot due to the corruption experienced. Clearly it seems to be the case that "rewrite all files and hope for the best" seems to be better than nothing. I just want to know if there are better ways so that I can document this as a recovery procedure. Would an fsck help? Or is it best to mkfs over?

Depends on the situation. If the stub is still alive, we can make it set the flag for an OS-level recovery and trigger a completely different procedure. See for the logic: https://uefi.org/specs/UEFI/2.10/03_Boot_Manager.html#boot-option-recovery 

But all of this is vapourware, I am planning to submit a more focused work group on how to make NixOS recover from broken boots in general.

Reply:
> (Side note: The comment in the atomic_copy function should then be changed as well. And I now noticed that its implementation is incorrect, since the sync is missing, but I'm not sure how much that changes on FAT after what you have written above.)

We already discussed this many times with @nikstur ; the rationale is that we cannot know under the hood that `rename` is not really POSIX.

So it's true that we are calling a POSIX rename that is supposed to be atomic, but, it will **never be** atomic because FAT32 as an ESP is simply a very bad idea, alas.

Reply:
> > How does this failure usually manifest itself in the cases you have experienced (truncated files, garbage content, …)? I'd like to be able to document this properly.
> 
> I didn't dive into failures related to FAT32 brokenness, but apparently, out there, there are in-the-wild broken implementations of FAT32 who does not know how to behave with the dirty flag set and does wrong things.
> 
> So globally, IME, anything can happen. Personally, I have seen truncated files mostly.

I don't care about "in-the-wild broken implementations of FAT32" here. The recovery system is presumably a NixOS ISO, so it will use a somewhat recent Linux kernel. I'm still talking about this situation:

> The only concern I have for not overwriting files is that repair guidelines are unclear now.

Since this is the case where power has failed and the user is now stuck with an unbootable system, I think it is fine to handle this with documentation. Previously the recovery instructions have apparently been "reinstall the bootloader [which will rewrite all the files] and hope for the best", not that I found this written down somewhere. I just want to know what's the best thing to write when I write something anyway. Which one of the following do you think would be better:

- `rm -fr <ESP>/Linux/nixos-generation-*.efi <ESP>/nixos`, then reinstall bootloader (roughly what happened before in the end result)
- `fsck` first, then above.
- `mkfs` over, then reinstall the bootloader to create the ESP from scratch.

Reply:
> > (Side note: The comment in the atomic_copy function should then be changed as well. And I now noticed that its implementation is incorrect, since the sync is missing, but I'm not sure how much that changes on FAT after what you have written above.)
> 
> We already discussed this many times with @nikstur ; the rationale is that we cannot know under the hood that `rename` is not really POSIX.

My point is that the comment above `atomic_copy` literally says "This is atomic, because a rename is an atomic operation on POSIX platforms.". This, from what I gather from what you wrote above, is just a lie when the function is called on a FAT32 filesystem (which it in fact is in nearly all cases).

> So it's true that we are calling a POSIX rename that is supposed to be atomic, but, it will **never be** atomic because FAT32 as an ESP is simply a very bad idea, alas.

Atomic write and atomic rename are somewhat related, but not really the same. Atomic rename to me means that after `rename`, either the old or the new file are there. Atomic write means that the target file is either present with correct content, or not present at all. This is implemented using "write temp, sync, rename" (of which the middle step is currently missing), and only requires much weaker guarantees from `rename`: it is fine if the file is present at both places, or even absent entirely.

Reply:
> I don't care about "in-the-wild broken implementations of FAT32" here. The recovery system is presumably a NixOS ISO, so it will use a somewhat recent Linux kernel. I'm still talking about this situation:

> Since this is the case where power has failed and the user is now stuck with an unbootable system, I think it is fine to handle this with documentation. Previously the recovery instructions have apparently been "reinstall the bootloader [which will rewrite all the files] and hope for the best", not that I found this written down somewhere. I just want to know what's the best thing to write when I write something anyway. Which one of the following do you think would be better:

Yes, but both of them are related… We cannot assume that in-the-wild implementations are good and we have to document this properly, for the same reason, everyone implements workarounds in the kernel, in systemd-boot and even us! We have a workaround for broken FAT32 implementation in uefi-rs we explicitly reported.

> rm -fr <ESP>/Linux/nixos-generation-*.efi <ESP>/nixos, then reinstall bootloader (roughly what happened before in 
the end result)
> fsck first, then above.
> mkfs over, then reinstall the bootloader to create the ESP from scratch.

It is hard:

(1) on its own is not enough.

(2) on its own can fail and put the user in troubles with no bootables.

(3) offers the most guarantee of getting back in a normal state but you may kill other bootables, e.g. Windows, Dell recovery, etc. which the user cares about because this is the whole problem about modern computers and boot: we assume most of the time — single-boot — systems. It is not surprising we have users with dual-boot or multi-boot alas.

So (2) is probably the closest to OK.

Reply:
> My point is that the comment above atomic_copy literally says "This is atomic, because a rename is an atomic operation on POSIX platforms.". This, from what I gather from what you wrote above, is just a lie when the function is called on a FAT32 filesystem (which it in fact is in nearly all cases).

> Atomic write and atomic rename are somewhat related, but not really the same. Atomic rename to me means that after rename, either the old or the new file are there. Atomic write means that the target file is either present with correct content, or not present at all. This is implemented using "write temp, sync, rename" (of which the middle step is currently missing), and only requires much weaker guarantees from rename: it is fine if the file is present at both places, or even absent entirely.

Right, but what is your concrete proposal? We can just add a comment `// Actually, this is not true because FAT32 is not journaled, so if you are in the middle of the VFS rename kernel call and power disappears, you reboot into a drity FAT32 filesystem but there's no way to recover the data and the structure is in whatever state.` ?